format/tzx/tzx.go

```diff
@@ -0,0 +1,655 @@
+package tzx
+
+// https://worldofspectrum.net/TZXformat.html
+
+import (
+	"embed"
+
+	"golang.org/x/text/encoding/charmap"
+
+	"github.com/wader/fq/format"
+	"github.com/wader/fq/pkg/decode"
+	"github.com/wader/fq/pkg/interp"
+	"github.com/wader/fq/pkg/scalar"
+)
+
+//go:embed tzx.md
+var tzxFS embed.FS
+
+var tapFormat decode.Group
+
+func init() {
+	interp.RegisterFormat(
+		format.TZX,
+		&decode.Format{
+			Description: "TZX tape format for ZX Spectrum computers",
+			Groups:      []*decode.Group{format.Probe},
+			DecodeFn:    tzxDecode,
+			Dependencies: []decode.Dependency{
+				{Groups: []*decode.Group{format.TAP}, Out: &tapFormat},
+			},
+		})
+	interp.RegisterFS(tzxFS)
+}
+
+func tzxDecode(d *decode.D) any {
+	d.Endian = decode.LittleEndian
+
+	d.FieldRawLen("signature", 8*8, d.AssertBitBuf([]byte("ZXTape!\x1A")))
+	d.FieldU8("major_version")
+	d.FieldU8("minor_version")
+	decodeBlocks(d)
+
+	return nil
+}
+
+func decodeBlocks(d *decode.D) {
+	d.FieldArray("blocks", func(d *decode.D) {
+		for !d.End() {
+			d.FieldStruct("block", func(d *decode.D) {
+				decodeBlock(d)
+			})
+		}
+	})
+}
+
+func decodeBlock(d *decode.D) {
+	blocks := map[uint64]func(d *decode.D){
+		// ID: 10h (16d) | Standard Speed Data
+		// This block is replayed with the standard Spectrum ROM timing values
+		// (the values in curly brackets in block ID 11). The pilot tone
+		// consists of 8063 pulses if the first data byte (the flag byte)
+		// is < 128, 3223 otherwise.
+		0x10: func(d *decode.D) {
+			// Pause after this block (ms.) {1000}
+			d.FieldU16("pause")
+
+			// A single TAP Data Block
+			peekBytes := d.PeekBytes(2)                              // get the TAP data block length
+			length := uint16(peekBytes[1])<<8 | uint16(peekBytes[0]) // bytes are stored in LittleEndian
+			length += 2                                              // include the two bytes for this value
```

Comment:
Is the `ReadOneBlock` option or could the length*8 limit and the end check be enough?

Reply:
Right! In retrospect it even makes sense for these embedded tap blocks would be an array. As the TAP spec states: "a TAP file is simply one datablock or a group of 2 or more datablocks".